README.md
# ![Mochimo header](https://mochimo.org/wp-content/uploads/2018/06/mochimo-pq-logo.svg)
![Made with C](https://img.shields.io/badge/Made%20with-C-%23.svg?color=%230059ff&style=for-the-badge)
![Release](https://img.shields.io/github/release/mochimodev/mochimo.svg?color=%230059ff&style=for-the-badge)
[![License](https://img.shields.io/badge/license-Modified%20MPL%202.0-%23.svg?color=%230059ff&style=for-the-badge)](LICENSE.PDF)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?color=%230059ff&style=for-the-badge)](http://makeapullrequest.com)  
<sup>*You must read and agree to the [LICENSE.PDF](LICENSE.PDF) file prior to running our code.*</sup>

### Mochimo Project - Main Net Development Repository
The Mochimo Crypto Engine Main Net code is a fully functional crypto currency network node with text-based developer's wallet included. The code will compile without issue on most Linux-based gcc compatible machines with a 64-bit long. However please be aware that developer support for troubleshooting will be restricted to those users running the recommended configuration.

## Usage
Thorough instructions for configuring your system and compiling can be found on the [Mochimo Wiki](http://www.mochiwiki.com)

## Recommended Hardware/Software
The included code is designed to be run on Ubuntu 16.04 LTS 64-Bit Desktop, with at least 2 processor threads and 2GB of RAM. At present the code is designed to function on NVIDIA GPUs, so both gcc and nvcc installations are required.

## JOIN THE COMMUNITY!
[![Discord](https://img.shields.io/badge/Mochimo%20Official%20Discord-%23.svg?color=%23505050&style=popout&logo=discord)](https://discord.gg/EHFqS5s)&emsp;
[![Twitter](https://img.shields.io/badge/@mochimocrypto-%23.svg?color=%23505050&style=popout&logo=twitter)](https://twitter.com/mochimocrypto)&emsp;
[![Reddit](https://img.shields.io/badge/r/Mochimo-%23.svg?color=%23505050&style=popout&logo=reddit)](https://www.reddit.com/r/mochimo/)  
Discord is our most active social forum where you can discuss Mochimo with the rest of the developer and beta testing community, but we are also active on other forms of social media.

## License
<sub>**The license to use versions of the code prior to v2.0 expired on December 31st, 2018. Use of the old code is strictly prohibited.**</sub>  
The current version of the code is released under an MPL2.0 derivative Open Source license.  
The community is free to develop and change the code with the caveat that any changes must be for the benefit of the Mochimo cryptocurrency network (with a number of exclusions).  
Please read [LICENSE.PDF](LICENSE.PDF) for more details on limitations and restrictions.

The Mochimo Main Net Package is copyright 2019 Adequate Systems, LLC.  
Please read the license file in the package for additional restrictions.

Contact: support@mochimo.org

CHANGES.TXT
/*
 * Copyright (c) 2019 by Adequate Systems, LLC.  All Rights Reserved.
 * See LICENSE.PDF   **** NO WARRANTY ****
 *
 */

Patch Level 37 (Mochimo v2.4.1)

      Changed PATCHLEVEL to 37 in mochimo.c and minertest.c
      Changed VERSIONSTR to "37" in mochimo.c and minertest.c
      Fixed previous hash check in checkproof() in proof.c
      Added byte Insyncup;  /* non-zero when syncup() runs */ to data.c
      Added Insyncup to syncup() in syncup.c
      Added addrlen to le_find() in ledger.c
      Added Tagidx[] to tag.c
      Changed tag_find() in tag.c, txclean.c, and mtxval.c
      Added tag_free() to tag.c, update.c
      Changed pval.c server.c, and config.h to allow 0xff pseudo-blocks
      Removed bigwait from server.c
      Added proper time_t vtime for vstart.lck check in server.c
      Added Insyncup to update() in update.c to help plog()'s
      Added char *solvestr to update() to reckon pushed, solved, and updated blocks.
      Simplified parameter logic in tag_valid() and calls from bval and txval.c.
      Added Syncup Function, Removing Contention
      made mochimo/bin/d/split
      modified bx.c to indicate tags not found in MTX.
      replaced 100 with NR_DST in bval.c mtxval.c txclean.c
      proof.c changed comments
      syncup.c changed fprintf's to plog's
      server.c changed ipltime from 600 to 300.
      rewrote checkproof() in proof.c
      changed comments and plog's in gettx.c
      fixed missing error return lines in bval2() in gettx.c:
      put back V23TRIGGER check in checkproof() in proof.c
      tried to improve comments and plog() messages.
      removed FILE *rlog from syncup().
      added send_found() to syncup.c.
      added Send found message to low weight peer in refresh_ipl()
         in renew.c.
      added MTXTRIGGER to bval.c mtxval.c txclean.c txval.c config.h.
      added support for third-party utility integration, with:
      added system call to init-external.sh on initial system sync
      added system call to update-external.sh on successful block update
      added advanced support for block-explorer export functionality
      Added support for third-party utility triggered system restart()
      Added directories mochimo/src/test and mochimo/src/old for testing this build
      Added weight checks to contention() and checkproof()
      Bugs fixed in checkproof()
      Added functions sub_weight(), and past_weight() to proof.c
      Added return code to send_found() in update.c
      Added sftimer for send_found() in server.c
      Added vstart.lck restart trigger (Verisimility) to server.c
      Updated comments
      Fixed bug in past_weight() to skip NG blocks
      Fixed bug in checkproof() to skip Difficulty check on init.
      Fixed bug in syncup() w/first previous NG block
      Swapped Bail(1) and Bail(2) to avoid recomputing past_weight() if
         first trailer doesn't match.
      Improved Code readability / comments for checkproof() bail conditions 6 & 7
      Fixed bug in syncup() to skip NG blocks

Patch Level 34
added FPGA-Resistant Algorithm
  -Extensible High-Memory Algo v24()
added Multi-Destination Transactions
  -Scales up high volume third-party payment systems.
added security fixes and networking tweaks 
 
Patch Level 33 (Mochimo Version 2.3)

added Pseudoblocks for mid-block difficulty adjustment
  -Impossible for blocks to exceed 15m49s after v2.3
added security fixes to MROOT creation
  -Removes certain spoofing attack vectors
added support for ZERO-tag OP_BAL Requests
  -Allows address lookup and balance query without TAG
  -Allows wallet recovery from seed phrase
added TXCLEAN Queue Re-validation following block updates
  -Prevents known cornercase block validation failures
  -Prevents all attack vectors involving poisoned TXs in the TXCLEAN queue
added low-balance pruning consensus mechanism (CAROUSEL)
  -Allows the community to clean low balances out of the ledger by consensus
  -Keeps the blockchain free of bloat, and recovers from spam attacks
added optional mining fee adjustment per miner
  -Allows future miners to create a mining fee market after mining rewards are gone
added TFILE PoPOW Chain in OP_FOUND
  -Allows nodes to definitively confirm an advertising node really solved a block
added Watchdog timer to restart and resync if no blocks solved or updated in 30 minutes +/- 10 minutes
  -Prevents Block of Death, Stuck at 0x0 events caused by temporary internet outage for nodes
  -Allows nodes to dynamically recover from any number of possible failure cases
added Upload Bandwidth limit (default = 5MB/s Upload, user configurable)
  -Prevents certain kinds of spam attacks against nodes
added NOMINER feature, initialized with -n at runtime
  -Allows a node to run in relay mode only without mining blocks
added command line compile options for CPU or GPU, merging both development branches
  -Involed with ./makeunx bin -DCPU  -or-  ./makeunx bin -DCUDA
replaced inbound TX uniqueness test (CRC) with address-based validation
  -Allows the system to scale past 65,536 TXs per block at some indeterminate future date

Patch Level 32 (Mochimo Version 2.2)

added support for diverse nVidia GPU Models
CUDA Code optimizations for average HP/s +10-20%
improved sanity checking in get_eon (prep for ring signatures)
added OPCODE 15 & 16 (pull candidate / push solved blocks)
added support for Headless Miners and Mining Pools
adjusted wait time up to 300 seconds from 180 if no quorum found
fixed random seed issue in rand.c
added node capability bits to identify server capabilities during handshake
added new execute and dispatch functions for handling headless miner requests
added new reaping function for terminating stale child processes related to headless miners
various community requested patches


Patch Level 31 (Mochimo Version 2.0)

October 27th, 2018

CHANGELOG

added new open source license
removed default maddr.dat
update system version number to 2

src/bval.c
/* bval.c  Block Validator
 *
 * Copyright (c) 2019 by Adequate Systems, LLC.  All Rights Reserved.
 * See LICENSE.PDF   **** NO WARRANTY ****
 *
 * The Mochimo Project System Software
 *
 * Date: 8 January 2018
 *
 * NOTE: Invoked by server.c update() by wait on system()
 *
 * Returns exit code 0 on successful validation,
 *                   1 I/O errors, or
 *                   >= 2 Peerip needs pinklist().
 *
 * Inputs:  argv[1],    rblock.dat, the block to validate
 *          ledger.dat  the ledger of address balances
 *
 * Outputs: ltran.dat  transaction file to post against ledger.dat
 *          exit status 0=valid or non-zero=not valid.
 *          renames argv[1] to "vblock.dat" on good validation.
*/


#include "config.h"
#include "mochimo.h"
#define closesocket(_sd) close(_sd)
char *trigg_check(byte *in, byte d, byte *bnum);
void trigg_expand2(byte *in, char *out);

#define EXCLUDE_NODES   /* exclude Nodes[], ip, and socket data */
#include "data.c"

#include "error.c"
#include "rand.c"
#include "crypto/crc16.c"
#include "add64.c"
#include "util.c"
#include "daemon.c"
#include "ledger.c"

#define EXCLUDE_RESOLVE
#include "tag.c"
#include "algo/peach/peach.c"
#include "mtxval.c"  /* for mtx */

word32 Tnum = -1;    /* transaction sequence number */
char *Bvaldelfname;  /* set == argv[1] to delete input file on failure */
TXQENTRY *Q2;        /* tag mods */


void cleanup(int ecode)
{
   if(Q2 != NULL) free(Q2);
   unlink("ltran.tmp");
   if(Bvaldelfname) unlink(Bvaldelfname);
   if(Trace) plog("cleanup() with ecode %i", ecode);
   exit(1);  /* no pink-list */
}

void drop(char *message)
{
   if(Trace && message)
      plog("bval: drop(): %s TX index = %d", message, Tnum);
   cleanup(3);
}


void baddrop(char *message)
{
   if(Trace && message)
      plog("bval: baddrop(): %s from: %s  TX index = %d",
           message, ntoa((byte *) &Peerip), Tnum);
   /* add Peerip to epoch pink list */
   cleanup(3);  /* put on epink.lst */
}


void bail(char *message)
{
   if(message) error("bval: %s", message);
   cleanup(1);
}


#if ADDR_TAG_LEN != 12
   ADDR_TAG_LEN must be 12 for tag code in bval.c
#endif


/* Invocation: bval file_to_validate */
int main(int argc, char **argv)
{
   BHEADER bh;             /* fixed length block header */
   static BTRAILER bt;     /* block trailer */
   static TXQENTRY tx;     /* Holds one transaction in the array */
   FILE *fp;               /* to read block file */
   FILE *ltfp;             /* ledger transaction output file ltran.tmp */
   word32 hdrlen, tcount;  /* header length and transaction count */
   int cond;
   static LENTRY src_le;            /* source and change ledger entries */
   word32 total[2];                 /* for 64-bit maths */
   static byte mroot[HASHLEN];      /* computed Merkel root */
   static byte bhash[HASHLEN];      /* computed block hash */
   static byte tx_id[HASHLEN];      /* hash of transaction and signature */
   static byte prev_tx_id[HASHLEN]; /* to check sort */
   static SHA256_CTX bctx;  /* to hash entire block */
   static SHA256_CTX mctx;  /* to hash transaction array */
   word32 bnum[2], stemp;
   static word32 mfees[2], mreward[2];
   unsigned long blocklen;
   int count;
   static byte do_rename = 1;
   static byte pk2[WOTSSIGBYTES], message[32], rnd2[32];  /* for WOTS */
   static char *haiku;
   static char haikufull[256];
   word32 now;
   TXQENTRY *qp1, *qp2, *qlimit;   /* tag mods */
   clock_t ticks;
   static word32 tottrigger[2] = { V23TRIGGER, 0 };
   static word32 v24trigger[2] = { V24TRIGGER, 0 };
   MTX *mtx;
   static byte addr[TXADDRLEN];  /* for mtx scan 4 */
   int j;  /* mtx */
   static TXQENTRY txs;     /* for mtx sig check */


   ticks = clock();
   fix_signals();
   close_extra();

   if(argc < 2) {
      printf("\nusage: bval {rblock.dat | file_to_validate} [-n]\n"
             "  -n no rename, just create ltran.dat\n"
             "This program is spawned from server.c\n\n");
      exit(1);
   }

   if(sizeof(MTX) != sizeof(TXQENTRY)) bail("bad MTX size");

   if(argc > 2 && argv[2][0] == '-') {
      if(argv[2][1] == 'n') do_rename = 0;
   }

   if(strcmp(argv[1], "rblock.dat") == 0) Bvaldelfname = argv[1];
   unlink("vblock.dat");
   unlink("ltran.dat");

   /* get global block number, peer ip, etc. */
   if(read_global() != VEOK)
      bail("Cannot read_global()");

   if(Trace) Logfp = fopen(LOGFNAME, "a");

   /* open ledger read-only */
   if(le_open("ledger.dat", "rb") != VEOK)
      bail("Cannot open ledger.dat");

   /* create ledger transaction temp file */
   ltfp = fopen("ltran.tmp", "wb");
   if(ltfp == NULL) bail("Cannot create ltran.tmp");

   /* open the block to validate */
   fp = fopen(argv[1], "rb");
   if(!fp) {
badread:
      bail("Cannot read input rblock.dat");
   }
   if(fread(&hdrlen, 1, 4, fp) != 4) goto badread;  /* read header length */
   /* regular fixed size block header */
   if(hdrlen != sizeof(BHEADER))
      drop("bad hdrlen");

   /* compute block file length */
   if(fseek(fp, 0, SEEK_END)) goto badread;
   blocklen = ftell(fp);

   /* Read block trailer:
    * Check phash, bnum,
    * difficulty, Merkel Root, nonce, solve time, and block hash.
    */
   if(fseek(fp, -(sizeof(BTRAILER)), SEEK_END)) goto badread;
   if(fread(&bt, 1, sizeof(BTRAILER), fp) != sizeof(BTRAILER))
      drop("bad trailer read");
   if(cmp64(bt.mfee, Mfee) < 0)
      drop("bad mining fee");
   if(get32(bt.difficulty) != Difficulty)
      drop("difficulty mismatch");

   /* Check block times and block number. */
   stemp = get32(bt.stime);
   /* check for early block time */
   if(stemp <= Time0) drop("E");  /* unsigned time here */
   now = time(NULL);
   if(stemp > (now + BCONFREQ)) drop("F");
   add64(Cblocknum, One, bnum);
   if(memcmp(bnum, bt.bnum, 8) != 0) drop("bad block number");
   if(cmp64(bnum, tottrigger) > 0 && Cblocknum[0] != 0xfe) {
      if((word32) (stemp - get32(bt.time0)) > BRIDGE) drop("TOT");
   }

   if(memcmp(Cblockhash, bt.phash, HASHLEN) != 0)
      drop("previous hash mismatch");

   /* check enforced delay, collect haiku from block */
   if(cmp64(bnum, v24trigger) > 0) {
      if(peach(&bt, get32(bt.difficulty), NULL, 1)){
         drop("peach validation failed!");
      }

      trigg_expand2(bt.nonce, haikufull);
      if(!Bgflag) printf("\n%s\n\n", haikufull);
   }
   if(cmp64(bnum, v24trigger) <= 0) {
      if((haiku = trigg_check(bt.mroot, bt.difficulty[0], bt.bnum)) == NULL) {
      drop("trigg_check() failed!");
      }
      if(!Bgflag) printf("\n%s\n\n", haiku);
   }

   /* Read block header */
   if(fseek(fp, 0, SEEK_SET)) goto badread;
   if(fread(&bh, 1, hdrlen, fp) != hdrlen)
      drop("short header read");
   get_mreward(mreward, bnum);
   if(memcmp(bh.mreward, mreward, 8) != 0)
      drop("bad mining reward");
   if(HAS_TAG(bh.maddr))
      drop("bh.maddr has tag!");

   /* fp left at offset of Merkel Block Array--ready to fread() */

   sha256_init(&bctx);   /* begin entire block hash */
   sha256_update(&bctx, (byte *) &bh, hdrlen);  /* ... with the header */

   if(NEWYEAR(bt.bnum)) memcpy(&mctx, &bctx, sizeof(mctx));

   /*
    * Copy transaction count from block trailer and check.
    */
   tcount = get32(bt.tcount);
   if(tcount == 0 || tcount > MAXBLTX)
      baddrop("bad bt.tcount");
   if((hdrlen + sizeof(BTRAILER) + (tcount * sizeof(TXQENTRY))) != blocklen)
      drop("bad block length");

   /* temp TX tag processing queue */
   Q2 = malloc(tcount * sizeof(TXQENTRY));
   if(Q2 == NULL) bail("no memory!");

   /* Now ready to read transactions */
   if(!NEWYEAR(bt.bnum)) sha256_init(&mctx);   /* begin Merkel Array hash */

   /* Validate each transaction */
   for(Tnum = 0; Tnum < tcount; Tnum++) {
      if(Tnum >= MAXBLTX)
         drop("too many TX's");
      if(fread(&tx, 1, sizeof(TXQENTRY), fp) != sizeof(TXQENTRY))
         drop("bad TX read");
      if(memcmp(tx.src_addr, tx.chg_addr, TXADDRLEN) == 0)
         drop("src == chg");
      if(!ismtx(&tx) && memcmp(tx.src_addr, tx.dst_addr, TXADDRLEN) == 0)
         drop("src == dst");

      if(cmp64(tx.tx_fee, Mfee) < 0) drop("tx_fee is bad");

      /* running block hash */
      sha256_update(&bctx, (byte *) &tx, sizeof(TXQENTRY));
      /* running Merkel hash */
      sha256_update(&mctx, (byte *) &tx, sizeof(TXQENTRY));
      /* tx_id is hash of tx.src_add */
      sha256(tx.src_addr, TXADDRLEN, tx_id);
      if(memcmp(tx_id, tx.tx_id, HASHLEN) != 0)
         drop("bad TX_ID");

      /* Check that tx_id is sorted. */
      if(Tnum != 0) {
         cond = memcmp(tx_id, prev_tx_id, HASHLEN);
         if(cond < 0)  drop("TX_ID unsorted");
         if(cond == 0) drop("duplicate TX_ID");
      }
      /* remember this tx_id for next time */
      memcpy(prev_tx_id, tx_id, HASHLEN);

      /* check WTOS signature */
      if(ismtx(&tx) && get32(Cblocknum) >= MTXTRIGGER) {
         memcpy(&txs, &tx, sizeof(txs));
         mtx = (MTX *) &txs;
         memset(mtx->zeros, 0, NR_DZEROS);  /* always signed when zero */
         sha256(txs.src_addr, SIG_HASH_COUNT, message);
      } else {
         sha256(tx.src_addr, SIG_HASH_COUNT, message);
      }
      memcpy(rnd2, &tx.src_addr[TXSIGLEN+32], 32);  /* copy WOTS addr[] */
      wots_pk_from_sig(pk2, tx.tx_sig, message, &tx.src_addr[TXSIGLEN],
                       (word32 *) rnd2);
      if(memcmp(pk2, tx.src_addr, TXSIGLEN) != 0)
         baddrop("WOTS signature failed!");

      /* look up source address in ledger */
      if(le_find(tx.src_addr, &src_le, NULL, 0) == FALSE)
         drop("src_addr not in ledger");

      total[0] = total[1] = 0;
      /* use add64() to check for carry out */
      cond =  add64(tx.send_total, tx.change_total, total);
      cond += add64(tx.tx_fee, total, total);
      if(cond) drop("total overflow");

      if(cmp64(src_le.balance, total) != 0)
         drop("bad transaction total");
      if(!ismtx(&tx)) {
         if(tag_valid(tx.src_addr, tx.chg_addr, tx.dst_addr, bt.bnum)
            != VEOK) drop("tag not valid");
      } else {
         if(mtx_val((MTX *) &tx, Mfee) != 0) drop("bad mtx_val()");
      }

      memcpy(&Q2[Tnum], &tx, sizeof(TXQENTRY));  /* copy TX to tag queue */

      if(add64(mfees, tx.tx_fee, mfees)) {
fee_overflow:
         bail("mfees overflow");
      }
   }  /* end for Tnum */
   if(NEWYEAR(bt.bnum))
      /* phash, bnum, mfee, tcount, time0, difficulty */
      sha256_update(&mctx, (byte *) &bt, (HASHLEN+8+8+4+4+4));

   sha256_final(&mctx, mroot);  /* compute Merkel Root */
   if(memcmp(bt.mroot, mroot, HASHLEN) != 0)
      baddrop("bad Merkle root");

   sha256_update(&bctx, (byte *) &bt, sizeof(BTRAILER) - HASHLEN);
   sha256_final(&bctx, bhash);
   if(memcmp(bt.bhash, bhash, HASHLEN) != 0)
      drop("bad block hash");

   /* tag search  Begin ... */
   qlimit = &Q2[tcount];
   for(qp1 = Q2; qp1 < qlimit; qp1++) {
      if(!HAS_TAG(qp1->src_addr)
         || memcmp(ADDR_TAG_PTR(qp1->src_addr), ADDR_TAG_PTR(qp1->chg_addr),
                   ADDR_TAG_LEN) != 0) continue;
      /* Step 2: Start another big-O n squared, nested loop here... */
      for(qp2 = Q2; qp2 < qlimit; qp2++) {
         if(qp1 == qp2) continue;  /* added -trg */
         if(ismtx(qp2)) continue;  /* skip multi-dst's for now */
         /* if src1 == dst2, then copy chg1 to dst2 -- 32-bit for DSL -trg */
         if(   *((word32 *) ADDR_TAG_PTR(qp1->src_addr))
            == *((word32 *) ADDR_TAG_PTR(qp2->dst_addr))
            && *((word32 *) (ADDR_TAG_PTR(qp1->src_addr) + 4))
            == *((word32 *) (ADDR_TAG_PTR(qp2->dst_addr) + 4))
            && *((word32 *) (ADDR_TAG_PTR(qp1->src_addr) + 8))
            == *((word32 *) (ADDR_TAG_PTR(qp2->dst_addr) + 8)))
                   memcpy(qp2->dst_addr, qp1->chg_addr, TXADDRLEN);
      }  /* end for qp2 */
   }  /* end for qp1 */

   /* 
    * Three times is the charm...
    */
   for(Tnum = 0, qp1 = Q2; qp1 < qlimit; qp1++, Tnum++) {
      /* Re-do all the maths again... */
      total[0] = total[1] = 0;
      cond =  add64(qp1->send_total, qp1->change_total, total);
      cond += add64(qp1->tx_fee, total, total);
      if(cond) bail("scan3 total overflow");

      /* Write ledger transactions to ltran.tmp for all src and chg,
       * but only non-mtx dst
       * that will have to be sorted, read again, and applied by bup...
       */
      fwrite(qp1->src_addr,  1, TXADDRLEN, ltfp);
      fwrite("-",            1,         1, ltfp);  /* debit src addr */
      fwrite(total,          1,         8, ltfp);
      /* add to or create non-multi dst address */
      if(!ismtx(qp1) && !iszero(qp1->send_total, 8)) {
         fwrite(qp1->dst_addr,   1, TXADDRLEN, ltfp);
         fwrite("A",             1,         1, ltfp);
         fwrite(qp1->send_total, 1,         8, ltfp);
      }
      /* add to or create change address */
      if(!iszero(qp1->change_total, 8)) {
         fwrite(qp1->chg_addr,     1, TXADDRLEN, ltfp);
         fwrite("A",               1,         1, ltfp);
         fwrite(qp1->change_total, 1,         8, ltfp);
      }
   }  /* end for Tnum -- scan 3 */

   
   if(Tnum != tcount) bail("scan 3");
   /* mtx tag search  Begin scan 4 ...
    *
    * Write out the multi-dst trans using tag scan logic @
    * that more or less repeats the above big-O n-squared loops, and
    * expands the tags, and copies addresses around.
    */
   for(qp1 = Q2; qp1 < qlimit; qp1++) {
      if(!ismtx(qp1)) continue;  /* only multi-dst's this time */
      mtx = (MTX *) qp1;  /* poor man's union */
      /* For each dst[] tag... */
      for(j = 0; j < NR_DST; j++) {
         if(iszero(mtx->dst[j].tag, ADDR_TAG_LEN)) break; /* end of dst[] */
         memcpy(ADDR_TAG_PTR(addr), mtx->dst[j].tag, ADDR_TAG_LEN);
         /* If dst[j] tag not found, write money back to chg addr. */
         if(tag_find(addr, addr, NULL) != VEOK) {
            count =  fwrite(mtx->chg_addr, TXADDRLEN, 1, ltfp);
            count += fwrite("A", 1, 1, ltfp);
            count += fwrite(mtx->dst[j].amount, 8, 1, ltfp);
            if(count != 3) bail("bad I/O dst-->chg write");
            continue;  /* next dst[j] */
         }
         /* Start another big-O n-squared, nested loop here... scan 5 */
         for(qp2 = Q2; qp2 < qlimit; qp2++) {
            if(qp1 == qp2) continue;
            /* if dst[j] tag == any other src addr tag and chg addr tag,
             * copy other chg addr to dst[] addr.
             */
            if(!HAS_TAG(qp2->src_addr)) continue;
            if(memcmp(ADDR_TAG_PTR(qp2->src_addr),
                      ADDR_TAG_PTR(qp2->chg_addr), ADDR_TAG_LEN) != 0)
                         continue;
            if(memcmp(ADDR_TAG_PTR(qp2->src_addr), ADDR_TAG_PTR(addr), 
                      ADDR_TAG_LEN) == 0) {
                         memcpy(addr, qp2->chg_addr, TXADDRLEN);
                         break;
            }
         }  /* end for qp2 scan 5 */
         /* write out the dst transaction */
         count =  fwrite(addr, TXADDRLEN, 1, ltfp);
         count += fwrite("A", 1, 1, ltfp);
         count += fwrite(mtx->dst[j].amount, 8, 1, ltfp);
         if(count != 3) bail("bad I/O scan 4");
      }  /* end for j */
   }  /* end for qp1 */
   /* end mtx scan 4 */

   /* Create a transaction amount = mreward + mfees
    * address = bh.maddr
    */
   if(add64(mfees, mreward, mfees)) goto fee_overflow;
   /* Make ledger tran to add to or create mining address.
    * '...Money from nothing...'
    */
   count =  fwrite(bh.maddr, 1, TXADDRLEN, ltfp);
   count += fwrite("A",      1,         1, ltfp);
   count += fwrite(mfees,    1,         8, ltfp);
   if(count != (TXADDRLEN+1+8) || ferror(ltfp))
      bail("ltfp I/O error");

   free(Q2);  Q2 = NULL;
   le_close();
   fclose(ltfp);
   fclose(fp);
   rename("ltran.tmp", "ltran.dat");
   unlink("vblock.dat");
   if(do_rename)
      rename(argv[1], "vblock.dat");
   if(Trace)
      plog("bval: block validated to vblock.dat (%u usec.)",
           (word32) (clock() - ticks));
   if(argc > 2) printf("Validated\n");
   return 0;  /* success */
}  /* end main() */
